core/src/main/java/org/infinispan/persistence/async/AsyncNonBlockingStore.java

```diff
@@ -474,72 +441,67 @@ public void run() {
-   }
-
-   private void submitBatchIfNecessary() {
+      Maybe<CompletionStage<Void>> maybeRemove = Flowable.fromPublisher(removePublisher)
+            .flatMap(sp -> Flowable.fromPublisher(sp)
+                  .map(key -> submitModification(new RemoveModification(sp.getSegment(), key))), publisherCount)
+            .lastElement();
+      Maybe<CompletionStage<Void>> maybeModify = Flowable.fromPublisher(writePublisher)
+            .flatMap(sp -> Flowable.fromPublisher(sp)
+                  .map(me -> submitModification(new PutModification(sp.getSegment(), me))), publisherCount)
+            .lastElement();
+      return maybeRemove.mergeWith(maybeModify)
+            .lastStage(CompletableFutures.completedNull())
+            .thenCompose(java.util.function.Function.identity());
+   }
+
+   CompletionStage<Void> submitModification(Modification modification) {
+      CompletionStage<Void> nextBatchStage;
       boolean startNewBatch;
       synchronized (this) {
+         modification.apply(this);
+
          if (startNewBatch = batchFuture == null) {
             batchFuture = new CompletableFuture<>();
          }
+
+         if (pendingModifications.size() > modificationQueueSize) {
+            if (trace) {
+               log.tracef("Operation will not return immediately, must wait until current batch completes");
+            }
+            nextBatchStage = batchFuture.thenApplyAsync(CompletableFutures.toNullFunction(), nonBlockingExecutor);
+         } else {
+            nextBatchStage = CompletableFutures.completedNull();
+         }
       }
 
       if (startNewBatch) {
          if (trace) {
             log.tracef("Requesting a new async batch operation to be ran!");
          }
          // Any old object will work
-         requestFlowable.onNext(requestFlowable);
+         submitTask();
       }
+      return nextBatchStage;
    }
 
-   private synchronized CompletionStage<Void> asyncOrThrottledStage() {
-      if (pendingModifications.size() > modificationQueueSize) {
-         if (trace) {
-            log.tracef("Operation will not return immediately, must wait until current batch completes");
-         }
-         // We could have multiple waiting on the stage, so make sure we can don't block the thread
-         // that completes the stage
-         return batchFuture.thenApplyAsync(CompletableFutures.toNullFunction(), nonBlockingExecutor);
-      }
-      return CompletableFutures.completedNull();
+   @Override
+   public CompletionStage<Void> write(int segment, MarshallableEntry<? extends K, ? extends V> entry) {
+      assertNotStopped();
+      return submitModification(new PutModification(segment, entry));
+   }
+
+   @Override
+   public CompletionStage<Boolean> delete(int segment, Object key) {
+      assertNotStopped();
+      return submitModification(new RemoveModification(segment, key))
+            // We always assume it was removed with async
```

Comment:
It is used by stuff like `Map.remove` and stats. It is just that when the store is async we essentially make it optional for performance. The prior async store did the same thing.  https://github.com/infinispan/infinispan/blob/10.1.x/core/src/main/java/org/infinispan/persistence/async/AsyncCacheWriter.java#L233

Reply:
> @wburns I forgot to ask: could we track what the biggest batch was for an async store, and check that it's not much bigger than `modificationQueueSize` in the stress test?

I did that via tracing in the stress test when I verified the changes.

Reply:
I agree, but I am guessing we may want a stat for removes at some point. Either way there is no way the async store can properly update it the way it is used in CacheWriterInterceptor.

Reply:
We already have the stat (removeHits). It's just that it's computed in `CacheMgmtInterceptor` based on the value that was loaded in the invocation context, not on the result of the store operation:

https://github.com/infinispan/infinispan/blob/2d53d0a1e59f2dc3335235f9dafb8e93e611c60e/core/src/main/java/org/infinispan/interceptors/impl/CacheMgmtInterceptor.java#L495

Reply:
Either way this seems like a discussion for another time. I really hate getting stuck in the weeds on something unrelated to the original PR making the PR wait on it. If you would I love for you to create a JIRA about changing the store SPI to not have that return value and we can discuss on Zulip or something.

Reply:
Sorry if I gave that impression, but I never meant to hold this PR, I was just asking a question because I didn't have a close look at this in the non-blocking SPI PR.

Anyway, I created https://issues.redhat.com/browse/ISPN-12638 to replace the Boolean with Void